keps/sig-scheduling/20190926-default-even-pod-spreading.md

```diff
@@ -0,0 +1,302 @@
+
+```yaml
+defaultTopologySpreadConstraints:
+  - maxSkew: 1
+    topologyKey: "kubernetes.io/hostname"
+    whenUnsatisfiable: ScheduleAnyway
+  - maxSkew: 1
+    topologyKey: "failure-domain.beta.kubernetes.io/zone"
+    whenUnsatisfiable: ScheduleAnyway
+```
+
+### How user stories are addressed
+
+Let's say we have a cluster that has a topology based on physical hosts and racks.
+Then, an operator can set the following scheduler configuration:
+
+```yaml
+apiVersion: componentconfig/v1alpha1
+defaultTopologySpreadConstraints:
+  - maxSkew: 5
+    topologyKey: "example.com/topology/physical_host"
+    whenUnsatisfiable: ScheduleAnyway
+  - maxSkew: 15
+    topologyKey: "example.com/topology/rack"
+    whenUnsatisfiable: DoNotSchedule
+```
+
+Then, a workload author could have the following `ReplicaSet`:
+
+```yaml
+apiVersion: apps/v1
+kind: ReplicaSet
+metadata:
+  name: replicated_demo
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: demo
+  template:
+    metadata:
+      labels:
+        app: demo
+    spec:
+      containers:
+      - name: php-redis
+        image: example.com/registry/demo:latest
+```
+
+Note that the workload author didn't provide spreading constraints in the `pod.spec`.
+The following spreading constraints will be derived from the constraints defined in ComponentConfig,
+and will be applied at runtime:
+
+```yaml
+topologySpreadConstraints:
+  - maxSkew: 5
+    TopologyKey: "example.com/topology/physical_host"
+    WhenUnsatisfiable: ScheduleAnyway
+    selector:
+      matchLabels:
+        app: demo
+  - maxSkew: 15
+    TopologyKey: "example.com/topology/rack"
+    WhenUnsatisfiable: DoNotSchedule
+    selector:
+      matchLabels:
+        app: demo
+```
+
+Please note that these constraints are honored internally in the scheduler, but they are NOT
+persisted in the PodSpec via API Server.
+
+### Implementation Details
+
+#### In the metadata/predicates/priorities flow
+
+1. Calculate the spreading constraints for the pod as part of the metadata calculation.
+   Use the constraints provided by the pod or calculate the default ones if they don't provide any.
+1. When running the predicates or priorities, use the constraints stored in the metadata.
+
+#### In the scheduler framework
+
+1. Calculate spreading constraints for the pod in the `PreFilter` extension point. Store them
+   in the `PluginContext`.
+1. In the `Filter` and `Score` extension points, use the stored spreading constraints instead of
+   the ones defined by the pod.
+
+### Test Plan
+
+To ensure this feature to be rolled out in high quality. Following tests are mandatory:
+
+- **Unit Tests**: All core changes must be covered by unit tests.
+- **Integration Tests**: One integration test for the default rules and one for custom rules.
+- **Benchmark Tests**: A benchmark test that compare the default rules against `SelectorSpreadingPriority`.
+  The performance should be as close as possible.
```

Comment:
Sorry for jumping late into this discussion - but do we believe we can do that efficiently?
I'm asking in the context of large clusters - i.e. 5k nodes.

I want to ensure that we won't introduce a visible regression, because this may become a blocker.

Reply:
I think we should be able to do this as efficiently as the current logic, I don't see why we can't. I agree that it is a blocker if this will be causing performance regressions though.

Reply:
`SelectorSpreadingPriority` is currently part of the default algorithms that run. Theoretically, we are talking about the same problem, and then it should be possible to devise an algorithm for `EvenPodsSpreadPriority` with the same complexity.
We will be backing it with performance measurements, as seen in the Test Plan.

Reply:
@alculquicondor  perhaps we should explicitly call that out in the KEP?

Reply:
The Test Plan is part of the graduation criteria.